# Benchmark Summary Statistics
---

Description:

    This notebook performs the summary statisticks results for all benchmark functions (one at a time).
    The optimization process is repeated N=50 times (with default settings) and then we collect the results.
    
---

### Import python libraries and set up the project directory

In [23]:
import os, sys
import numpy as np

PROJECT_DIR = os.path.abspath('..')
sys.path.append(PROJECT_DIR)

### Import the custom PSO code

In [24]:
from star_pso.population.swarm import Swarm
from star_pso.population.particle import Particle
from star_pso.engines.standard_pso import StandardPSO
from star_pso.utils.auxiliary import cost_function

### Load all the Benchmark problems

In [25]:
# Dimensions: 1D
from star_pso.benchmarks.equal_maxima import EqualMaxima
from star_pso.benchmarks.five_uneven_peak_trap import FiveUnevenPeakTrap
from star_pso.benchmarks.uneven_decreasing_maxima import UnevenDecreasingMaxima

# Dimensions: 2D
from star_pso.benchmarks.himmelblau import Himmelblau
from star_pso.benchmarks.six_hump_camel_back import SixHumpCamelBack

# Dimensions: N >= 2D
from star_pso.benchmarks.shubert import Shubert
from star_pso.benchmarks.vincent import Vincent
from star_pso.benchmarks.rastrigin import Rastrigin

In [26]:
# Select one test problem.
benchmark = EqualMaxima()

### Define the cost function

Note: All the optimization problems are solved as maximizations.

In [27]:
@cost_function(minimize=False)
def func_benchmark(x_array: np.ndarray, **kwargs) -> float:
    # Return the solution.
    return benchmark.func(x_array).item()
# _end_def_

### Setup the experiment

In [28]:
# Number of repeatitions.
n_repeats = 50

# Swarm population size.
POP_SIZE = 100

# Maximum function evaluations.
MAX_F_EVALS = 50_000

# Store the results.
results = n_repeats * [0]

# Display Info.
print(f"Testing function: {benchmark.name}\n")

# Go time ...
for i in range(n_repeats):

    # Display current iteration.
    print(f"Running optimization: {i+1} out of {n_repeats} ... ", flush=True)

    # Get an initial (new) sample.
    x_sample = benchmark.sample_random_positions(n_pos=POP_SIZE)

    # Initial swarm population.
    swarm_t0 = Swarm([Particle(x) for x in x_sample])
    
    # Create the StandardPSO object that will carry on the optimization.
    test_PSO = StandardPSO(initial_swarm = swarm_t0, obj_func = func_benchmark,
                           x_min = benchmark.x_min, x_max = benchmark.x_max)
    # Run the optimization.
    test_PSO.run(max_it = 800, f_max_eval = MAX_F_EVALS, verbose = False,
                 options = {"w0": 0.70, "c1": 1.50, "c2": 1.50, "mode": "multimodal"})
    
    # Obtain the number of the global optima that were found for epsilon = 1.0e-4.
    gopt_found, gopt_total = benchmark.search_for_optima(test_PSO.swarm.best_n(POP_SIZE))

    # Update the results.
    results[i] = gopt_found

    # Blank line to declatter the screen.
    print(" ")
# _end_for_

print("Done.")

Testing function: Vincent_2D

Running optimization: 1 out of 50 ... 


11/29/2025 11:42:47 INFO: Initial f_optimal = 0.9975
11/29/2025 11:43:11 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 24.666 seconds.
 
Running optimization: 2 out of 50 ... 


11/29/2025 11:43:12 INFO: Initial f_optimal = 0.9955
11/29/2025 11:43:41 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 29.037 seconds.
 
Running optimization: 3 out of 50 ... 


11/29/2025 11:43:41 INFO: Initial f_optimal = 0.9993
11/29/2025 11:44:07 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 26.388 seconds.
 
Running optimization: 4 out of 50 ... 


11/29/2025 11:44:07 INFO: Initial f_optimal = 0.9963
11/29/2025 11:44:40 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 32.718 seconds.
 
Running optimization: 5 out of 50 ... 


11/29/2025 11:44:40 INFO: Initial f_optimal = 0.9999
11/29/2025 11:45:03 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.860 seconds.
 
Running optimization: 6 out of 50 ... 


11/29/2025 11:45:03 INFO: Initial f_optimal = 0.9980
11/29/2025 11:45:28 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 25.093 seconds.
 
Running optimization: 7 out of 50 ... 


11/29/2025 11:45:28 INFO: Initial f_optimal = 0.9956
11/29/2025 11:45:49 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.553 seconds.
 
Running optimization: 8 out of 50 ... 


11/29/2025 11:45:49 INFO: Initial f_optimal = 0.9896
11/29/2025 11:46:10 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.453 seconds.
 
Running optimization: 9 out of 50 ... 


11/29/2025 11:46:10 INFO: Initial f_optimal = 0.9999
11/29/2025 11:46:31 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.538 seconds.
 
Running optimization: 10 out of 50 ... 


11/29/2025 11:46:31 INFO: Initial f_optimal = 0.9986
11/29/2025 11:46:51 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.579 seconds.
 
Running optimization: 11 out of 50 ... 


11/29/2025 11:46:51 INFO: Initial f_optimal = 0.9989
11/29/2025 11:47:14 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.856 seconds.
 
Running optimization: 12 out of 50 ... 


11/29/2025 11:47:14 INFO: Initial f_optimal = 0.9993
11/29/2025 11:47:37 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.347 seconds.
 
Running optimization: 13 out of 50 ... 


11/29/2025 11:47:37 INFO: Initial f_optimal = 0.9991
11/29/2025 11:47:58 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.145 seconds.
 
Running optimization: 14 out of 50 ... 


11/29/2025 11:47:58 INFO: Initial f_optimal = 0.9953
11/29/2025 11:48:19 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.937 seconds.
 
Running optimization: 15 out of 50 ... 


11/29/2025 11:48:19 INFO: Initial f_optimal = 0.9939
11/29/2025 11:48:40 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.783 seconds.
 
Running optimization: 16 out of 50 ... 


11/29/2025 11:48:40 INFO: Initial f_optimal = 0.9988
11/29/2025 11:49:03 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.178 seconds.
 
Running optimization: 17 out of 50 ... 


11/29/2025 11:49:03 INFO: Initial f_optimal = 0.9994
11/29/2025 11:49:24 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.966 seconds.
 
Running optimization: 18 out of 50 ... 


11/29/2025 11:49:24 INFO: Initial f_optimal = 0.9995
11/29/2025 11:49:45 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.818 seconds.
 
Running optimization: 19 out of 50 ... 


11/29/2025 11:49:45 INFO: Initial f_optimal = 0.9998
11/29/2025 11:50:06 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.872 seconds.
 
Running optimization: 20 out of 50 ... 


11/29/2025 11:50:06 INFO: Initial f_optimal = 0.9979
11/29/2025 11:50:27 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.520 seconds.
 
Running optimization: 21 out of 50 ... 


11/29/2025 11:50:27 INFO: Initial f_optimal = 0.9972
11/29/2025 11:50:49 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.555 seconds.
 
Running optimization: 22 out of 50 ... 


11/29/2025 11:50:49 INFO: Initial f_optimal = 0.9988
11/29/2025 11:51:12 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.803 seconds.
 
Running optimization: 23 out of 50 ... 


11/29/2025 11:51:12 INFO: Initial f_optimal = 0.9971
11/29/2025 11:51:34 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.753 seconds.
 
Running optimization: 24 out of 50 ... 


11/29/2025 11:51:34 INFO: Initial f_optimal = 0.9970
11/29/2025 11:51:54 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 20.909 seconds.
 
Running optimization: 25 out of 50 ... 


11/29/2025 11:51:55 INFO: Initial f_optimal = 0.9979
11/29/2025 11:52:16 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.059 seconds.
 
Running optimization: 26 out of 50 ... 


11/29/2025 11:52:16 INFO: Initial f_optimal = 0.9997
11/29/2025 11:52:37 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.158 seconds.
 
Running optimization: 27 out of 50 ... 


11/29/2025 11:52:37 INFO: Initial f_optimal = 0.9980
11/29/2025 11:52:59 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.217 seconds.
 
Running optimization: 28 out of 50 ... 


11/29/2025 11:52:59 INFO: Initial f_optimal = 0.9976
11/29/2025 11:53:21 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.019 seconds.
 
Running optimization: 29 out of 50 ... 


11/29/2025 11:53:21 INFO: Initial f_optimal = 0.9987
11/29/2025 11:53:43 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.073 seconds.
 
Running optimization: 30 out of 50 ... 


11/29/2025 11:53:43 INFO: Initial f_optimal = 0.9998
11/29/2025 11:54:05 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.237 seconds.
 
Running optimization: 31 out of 50 ... 


11/29/2025 11:54:05 INFO: Initial f_optimal = 0.9975
11/29/2025 11:54:26 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.580 seconds.
 
Running optimization: 32 out of 50 ... 


11/29/2025 11:54:26 INFO: Initial f_optimal = 0.9983
11/29/2025 11:54:48 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 21.575 seconds.
 
Running optimization: 33 out of 50 ... 


11/29/2025 11:54:48 INFO: Initial f_optimal = 0.9992
11/29/2025 11:55:11 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.401 seconds.
 
Running optimization: 34 out of 50 ... 


11/29/2025 11:55:11 INFO: Initial f_optimal = 1.0000
11/29/2025 11:55:34 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 22.891 seconds.
 
Running optimization: 35 out of 50 ... 


11/29/2025 11:55:34 INFO: Initial f_optimal = 0.9994
11/29/2025 11:55:58 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.847 seconds.
 
Running optimization: 36 out of 50 ... 


11/29/2025 11:55:58 INFO: Initial f_optimal = 0.9971
11/29/2025 11:56:24 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 25.975 seconds.
 
Running optimization: 37 out of 50 ... 


11/29/2025 11:56:24 INFO: Initial f_optimal = 0.9992
11/29/2025 11:56:50 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 25.889 seconds.
 
Running optimization: 38 out of 50 ... 


11/29/2025 11:56:50 INFO: Initial f_optimal = 0.9997
11/29/2025 11:57:16 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 26.170 seconds.
 
Running optimization: 39 out of 50 ... 


11/29/2025 11:57:16 INFO: Initial f_optimal = 0.9991
11/29/2025 11:57:44 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 27.666 seconds.
 
Running optimization: 40 out of 50 ... 


11/29/2025 11:57:44 INFO: Initial f_optimal = 0.9994
11/29/2025 11:58:12 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 27.558 seconds.
 
Running optimization: 41 out of 50 ... 


11/29/2025 11:58:12 INFO: Initial f_optimal = 0.9999
11/29/2025 11:58:36 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.962 seconds.
 
Running optimization: 42 out of 50 ... 


11/29/2025 11:58:36 INFO: Initial f_optimal = 0.9984
11/29/2025 11:59:00 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.884 seconds.
 
Running optimization: 43 out of 50 ... 


11/29/2025 11:59:00 INFO: Initial f_optimal = 0.9991
11/29/2025 11:59:25 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 25.540 seconds.
 
Running optimization: 44 out of 50 ... 


11/29/2025 11:59:25 INFO: Initial f_optimal = 0.9994
11/29/2025 11:59:49 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.867 seconds.
 
Running optimization: 45 out of 50 ... 


11/29/2025 11:59:49 INFO: Initial f_optimal = 0.9996
11/29/2025 12:00:13 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.892 seconds.
 
Running optimization: 46 out of 50 ... 


11/29/2025 12:00:13 INFO: Initial f_optimal = 0.9991
11/29/2025 12:00:37 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 23.853 seconds.
 
Running optimization: 47 out of 50 ... 


11/29/2025 12:00:37 INFO: Initial f_optimal = 0.9991
11/29/2025 12:01:02 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 25.091 seconds.
 
Running optimization: 48 out of 50 ... 


11/29/2025 12:01:02 INFO: Initial f_optimal = 0.9979
11/29/2025 12:01:27 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 24.427 seconds.
 
Running optimization: 49 out of 50 ... 


11/29/2025 12:01:27 INFO: Initial f_optimal = 0.9993
11/29/2025 12:01:51 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 24.020 seconds.
 
Running optimization: 50 out of 50 ... 


11/29/2025 12:01:51 INFO: Initial f_optimal = 0.9995
11/29/2025 12:02:15 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 99


Final f_optimal = 1.0000
run: elapsed time = 24.663 seconds.
 
Done.


In [29]:
# Compute the "Peak Ratio".
peak_ratio = sum(results) / (gopt_total * n_repeats)

print(f"PR = {peak_ratio}")

PR = 0.5472222222222223


In [30]:
# Compute the "Success Rate".
success_rate = sum([r == gopt_total for r in results]) / n_repeats

print(f"SR = {success_rate}")

SR = 0.0


### Optimization Results

Accuracy level: $\epsilon = 1.0^{-4}$

| Benchmark Function | Dimensions | No. Global opt. | MaxFEs  | Population Size | Parameters | PR | SR |
|:---|:---|:---|:---|:---|:---|:---|:---|
| Equal Maxima | 1 | 5 | 50_000| 120 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Five Uneven Peak Trap | 1 | 2 | 50_000 | 120 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Uneven Decreasing Maxima | 1 | 1 | 50_000 | 85 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Himmelblau | 2 | 4 | 50_000 | 100 | (0.70, 1.50, 1.50) |1.0 | 1.0 |
| Six Hump Camel Back | 2 | 2 | 50_000 | 85 | (0.70, 1.50, 1.50) |1.0 | 1.0 |
| Rastrigin | 2 / 4| 6 / 12 | 100_000 / 200_000 | 150 / 400 | (0.70, 1.50, 1.50) |1.000 / 0.998 | 1.000 / 0.980 |
| Shubert | 2 / 3 | 18 / 81 | 200_000 / 400_000 | 600 / 1600 | (0.75, 1.50, 1.50) |0.983 / 0.755 | 0.740 / 0.000 |
| Vincent | 2 / 3 | 36 / 216 | 200_000 / 400_000 | 2000 / 4000 | (0.70, 1.50, 1.50) |0.744 / 0.392 | 0.000 / 0.000 |

### End of file